README.md
# Lightning In A Bot CLI
A tool for developing nodes on the Lightning In A Bot platform

## Installing
```
$ sudo npm install liab -g
```
## Requirements
- Node v >= 4

## Documentation

There are 3 commands: create, init, and upload.

### Create

### Init

Running the init command creates the necessary resources on the server. When the command is run successfully from within a local node directory, it will return a node Id from the server and add it to the .env file. This Id is required in order to upload code updates.

### Upload

Runnin the upload command creates the necessary resources locally that are the send to the server. You may use this command is used to update nodes as well. Make sure there is a valid node ID in the config.json file along with valid inputs for updatable keys (code will be updated so long as changes are saved). An example config.json:

	{
  	"id": "56ac2407dbc0091100e9a524",
 		"expressions": ["An expression", "another expression"],
 		"description": "This is a description",
 		"keys": [
      {"name":"count"},
      {
        "name":"datetime",
        "type":"datetime"
      },
      {
        "name":"is_plural",
        "type":"boolean"
      }
    ]
	}

##### Warning: DO NOT change ID during upload. It will fail to correctly update your node.




CHANGELOG.md
Change Log
- - -
## v0.6.0 (7/3/18)
-- liab-upload no longer updates node's config attributes, and only updates node's code
-- use of util.inspect in liab-test fixed to display complete result object
## v0.5.1 (4/24/17)
-- liab-test will now utilizes util.inspect to display deep objects
## v0.5.0 (4/12/17)
-- liab-upload command will now update the node name if it has changed
## v0.4.18 (01/25/17)
-- Updating parsing of config flags to use if ('some_key' in ProjectConfigs) to test if key exists rather than check if object populated.
## v0.4.17 (01/25/17)
-- error message update and parsing report_buidler boolean flag from config.json
## v0.4.16 (9/27/16)
-- minor fixes for liab-init and liab-upload
## v0.4.15 (9/26/16)
-- devConfig created only when using liab init -e dev and not in liab create. liab init for a production environment does not create the file
## v0.4.14 (9/26/16)
- development id's for creating, initializing, and uploading nodes in development environment obtained through devCongif.json file which will now be ignored in liabnodes.
## v0.4.13 (8/19/16)
- only index.js, package.json, and config.json checked at top level of node directory. all other files (including filters.js) must be placed in lib directory of node.
## v0.4.12 (8/19/16)
-  New gulp task to add lib folder to dis/. Cleanup now will remove all files AND folders from dist/
## v0.4.11 (8/7/16)
- fix liab-upload success message. '/upload' is now returning an updated node object, rather than success message 
## v0.4.10 (8/2/16)
- testEvent now dynamically loads what is assigned in test.json.
## v0.4.9 (8/1/16)
- node domain names can now be added to project config files
## v0.4.8 (7/22/16)
- liab-upload will now look for a filters.js file to include in the dist
## v0.4.7 (5/31/16)
- Now sending package.json and config.json to LIAB. Minor fix for local dev uploading
## v0.4.6 (5/24/16)
- Fix: existing node properties won't be overwritten with null when uploading, if the properties exist on server but not in config.json
## v0.4.5 (5/24/16)
- Node description, expressions, and keys can now be added to config.json and updated on the server when "liab upload" is called.
## v0.4.4 (5/4/16)
- Added check for config.id in "init" command. Will prompt usage of "upload" command if Id exists.
## v0.4.3 (4/28/16)
- Updated exports.handler boilerplate code to reflect Lambda Node.js v4 runtime changes
## v0.4.2 (4/27/16)
- Added "host" property to "event" argument
## v0.4.1 (4/12/16)
- Support for new "callback" argument in event.handler (only Node.js v4 Lambda runtime)
## v0.4 (4/7/16)
- Removed "dotenv" dependency for projects
- Updated liab-test to look for test cases in test.json, rather than being passed through command line
- The test event in liab-test now contains a "bot" property to pass test bot objects into the node
- Added config.json to store project info
## v0.3.5 (3/17/16)
- Happy birthday to me. What'd I get? A timer in the test method!
## v0.3.4 (2/10/16)
- Improved handling of different environments; use -e CLI option to send request to localhost
- New package name round 2! Hello, LIAB!
## v0.3.3 (1/7/16)
- init command saves the node's id to projects .env file
- upload command now uses update /nodes/{NODE_ID}/upload route
## v0.3.2 (12/28/15)
- test command now works with promises
- new init command decouples creation of node on server with updating of node
## v0.3.1 (8/30/15)
- New package name! Hello, Nerul.
## v0.3.0 (8/9/15)
- Package.json repo url now sent to Elf on upload
## v0.2.2 (8/6/15)
- Added lib/ to array of files in package.json
## v0.2.1 (8/6/15)
- Fixed bug with upload command
## v0.2.0 (8/5/15)
- Added "open [name]" command


lib/utilities.js
/***************************************************
Utilities

PLEASE CHECK TO MAKE FUNCTIONS ARE PROVIDED REQUIRED 
DEPENDENCIES!

THANKS! 
****************************************************/

var os = require('os');
var fs = require('fs');
var mkdirp = require('mkdirp');
var readline = require('readline');
var path = require('path');


var _exit = process.exit;


module.exports = {

	/**
	 * URL friendly string
	 *
	 * @param {String} str
	 */
	slugify: function(str) {
	  return str.toString().toLowerCase()
	    .replace(/\s+/g, '-')        // Replace spaces with -
	    .replace(/[^\w\-]+/g, '')   // Remove all non-word chars
	    .replace(/\-\-+/g, '-')      // Replace multiple - with single -
	    .replace(/^-+/, '')          // Trim - from start of text
	    .replace(/-+$/, '');         // Trim - from end of text
	},


	/**
	 * Checks if an object is empty
	 *
	 * @param {obj} object
	 */
	isEmptyObj: function(obj) {
	  for(var prop in obj) {
	    if(obj.hasOwnProperty(prop)) 
	    	return false;
	  }
	  return true;
	},


	/**
	 * echo str > path.
	 *
	 * @param {String} path
	 * @param {String} str
	 */
	write: function(path, str, mode) {
	  fs.writeFileSync(path, str, { mode: mode || 0666 });
	  console.log('   \x1b[36mcreate\x1b[0m : ' + path);
	},


	/**
	 * Mkdir -p.
	 *
	 * @param {String} path
	 * @param {Function} fn
	 */
	mkdir: function(path, fn) {
	  mkdirp(path, 0755, function(err){
	    if (err) throw err;
	    console.log('   \033[36mcreate\033[0m : ' + path);
	    fn && fn();
	  });
	},


	/**
	 * Check if the given directory `path` is empty.
	 *
	 * @param {String} path
	 * @param {Function} fn
	 */
	emptyDirectory: function(path, fn) {
	  fs.readdir(path, function(err, files){
	    if (err && 'ENOENT' != err.code) throw err;
	    fn(!files || !files.length);
	  });
	},


	/**
	 * Determine if launched from cmd.exe
	 */
	launchedFromCmd: function() {
	  return process.platform === 'win32'
	    && process.env._ === undefined;
	},


	/**
	 * Load template file.
	 */
	loadTemplate: function(name) {
	  return fs.readFileSync(path.join(__dirname, '..', 'templates', name), 'utf-8');
	},


	/**
	 * Prompt for confirmation on STDOUT/STDIN
	 */

	confirm: function(msg, callback) {
	  var rl = readline.createInterface({
	    input: process.stdin,
	    output: process.stdout
	  });

	  rl.question(msg, function (input) {
	    rl.close();
	    callback(/^y|yes|ok|true$/i.test(input));
	  });
	},


	/**
	 * Graceful exit for async STDIO
	 */
	exit: function(code) {
	  // flush output for Node.js Windows pipe bug
	  // https://github.com/joyent/node/issues/6247 is just one bug example
	  // https://github.com/visionmedia/mocha/issues/333 has a good discussion
	  function done() {
	    if (!(draining--)) _exit(code);
	  }

	  var draining = 0;
	  var streams = [process.stdout, process.stderr];

	  module.exports.exit.exited = true;

	  streams.forEach(function(stream){
	    // submit empty write request and wait for completion
	    draining += 1;
	    stream.write('', done);
	  });

	  done();
	}
	
};

templates/gulpfile.js
// Deprecated file [7/31/15]

var gulp 				= require('gulp');
var gutil 			= require('gulp-util');
var del 				= require('del');
var install 		= require('gulp-install');
var zip 				= require('gulp-zip');
var request 		= require('request');
var fs 					= require('fs');
var runSequence = require('run-sequence');
var netrc 			= require('node-netrc');
var pkg		 			= require('./package.json');


var SITE_URL = 'http://localhost:3000';
var SITE_DOMAIN = 'localhost:3000';

 
gulp.task('clean', function(cb) {
  return del('./dist',
    del('./dist.zip', cb)
  );
});
 

gulp.task('js', function() {
  return gulp.src('index.js')
    .pipe(gulp.dest('dist/'))
});
 

gulp.task('npm', function() {
  return gulp.src('./package.json')
    .pipe(gulp.dest('./dist/'))
    .pipe(install({production: true}));
});
 

gulp.task('zip', function() {
  return gulp.src(['dist/**/*', '!dist/package.json', 'dist/.*'])
    .pipe(zip('dist.zip'))
    .pipe(gulp.dest('./'));
});
 

gulp.task('upload', function() {
	var data = {
		'deps' 								: Object.keys(pkg.dependencies),
		'index.js' 						: fs.createReadStream('./index.js')
	};
	if (pkg.description) { data.output_description = pkg.description; }
	var auth = netrc(SITE_DOMAIN);
	data[pkg.name + '.zip'] = fs.createReadStream('./dist.zip');
	return request({
		uri 			: SITE_URL + '/api/v1/i/' + pkg.name,
		method 		: 'PUT',
		formData 	: data,
		auth			: {
    	'user': auth.login,
    	'pass':  auth.password,
    	'sendImmediately': true
		}
	}, function (err, res, body) {
		if (err) { gutil.log(err); }
		gutil.log(JSON.parse(body).message);
	});
});


gulp.task('default', function(callback) {
  return runSequence(
    ['clean'],
    ['js', 'npm'],
    ['zip'],
    ['upload'],
    callback
  );
});